### PR TITLE
Validate acs url

### PR DESCRIPTION
**Pull request by jphenow**
Matching [standard](http://docs.oasis-open.org/security/saml/v2.0/saml-core-2.0-os.pdf) (line 2063) so that IDP performs validation on the AssertionConsumerServiceURL field to match Metadata Host or some other configured Host to prevent MitM attacks

### CODE REVIEW

**Comment by ikkisoft** [comment]
My understanding of the _Destination_ (saml 2 core lines 1477-1482) vs _AssertionConsumerServiceURL_ (saml 2 core lines 1483-1488) is that they're the same value, but  just used for different purposes:

- Destination is the URL of the endpoint where the Service Provider is expecting the response
- _AssertionConsumerServiceURL_ is the same URL within the IDP request, where peer IDP must send the response to

So, I think you can simply use the value you already have in the config for the service_provide and not the metadata URL.

**some-issuer-url.com/saml --> some-issuer-url.com**

```
service_providers = {
    "some-issuer-url.com/saml" => {
      fingerprint: "9E:65:2E:03:06:8D:80:F2:86:C7:6C:77:A1:D9:14:97:0A:4D:F4:4D",
      metadata_url: "http://some-issuer-url.com/saml/metadata"
    },
  }
```
If you allow override via config file (as implemented), I think is safe enough to enforce it since people can change the config after the new deployment. Especially, if it's clear in the README.

Btw, reading the standard (http://docs.oasis-open.org/security/saml/v2.0/saml-core-2.0-os.pdf line 2063) they are actually saying that the IDP is required to perform validation on the _AssertionConsumerServiceURL_ field:

```
The responder MUST ensure by some means that the value specified is in fact associated
with the requester.
```

**Comment by ikkisoft** [comment]
For this issue, you can use CVE-2018-18604 (in case you will send out security update notices, etc.)

**Comment by itstehkman** [comment]
I agree with @ikkisoft. I think this PR is mostly good. The only issue I see is if you've configured the IdP with N acceptable_response_hosts, where one of these could maliciously accept a response meant for another host:

EG: 

acceptable_response_hosts = [A, B]

I make a SAMLRequest for host A, and it gets maliciously modified such that its Destination / ACS URL is sent to B instead. In the unlikely case that B would accept the assertion, this is an issue.

I think it is simpler and more accurate to just have this on a per-service-provider configuration level without checking a set of acceptable hosts.

File: lib/saml_idp/request.rb [hunk 1]
```diff
@@ -136,6 +141,20 @@ def session_index
       @_session_index ||= xpath("//samlp:SessionIndex", samlp: samlp).first.try(:content)
     end
 
+    def acceptable_reponse_hosts
```
Comment (itstehkman) [inline lib/saml_idp/request.rb]: also small nit: typo -> acceptable_response_hosts

**Comment by itstehkman** [comment]
@jphenow is this ready? I think this implementation should suffice.

**Comment by jphenow** [comment]
Updated to set on per-service-provider basis with a list just for flexibility. If that's a security hole I deem that a hole on their configuration, not the lib. Let me know what you think and I can make a minor version release later today or tomorrow

**Comment by itstehkman** [comment]
@jphenow this looks great! Thanks for the updates. LGTM.

**Comment by jphenow** [comment]
Released as `0.8.0`

**Pull request merged**